### PR TITLE
Make scanner skip packages with paths excluded in .ort.yml

### PR DESCRIPTION
**Pull request by nnobelis**
Fix to ignore packages in scanner which are from a nested repository with a path that has been excluded in .ort.yml.
 
Please ignore the tests, they were quickly done for TDD.

Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

File: model/src/main/kotlin/OrtResult.kt [hunk 1]
```diff
@@ -143,8 +143,22 @@ data class OrtResult(
                 project.scopes.forEach { scope ->
                     val isScopeExcluded = getExcludes().isScopeExcluded(scope)
 
-                    val dependencies = scope.collectDependencies()
+                    val excludedDependencies = mutableSetOf<Identifier>()
+                    val dependencies = scope.collectDependencies(filterPredicate = { packageReference ->
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt:147]: Just use: `scope.collectDependencies { ... }`

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]
Could you explain the problem you are trying to fix with a bit more detail? Best with an example?

**Comment by nnobelis** [comment]
Yes, apologies. This PR indeed lacks proper description and I will try to explain the best I could.

I have a CPP project with a spdx.yml file. In this file, several packages are defined, some of which are dependencies of others :
```
- SPDXID: "SPDXRef-Package-AAA_Tools"
  copyrightText: "NONE"
  downloadLocation: "https://example.com/AAA_tools.git"
  filesAnalyzed: false
  homepage: "NONE"
  licenseConcluded: "NONE"
  licenseDeclared: "LicenseRef-Proprietary-AAA-Systems"
  packageFileName: "./external/AAA_tools"
  name: "AAA_Tools"
  originator: "Organization: AAA Systems"
```
and :

```
spdxElementId: "SPDXRef-Package-MyApplication"
  relationshipType: "DEPENDS_ON"
  relatedSpdxElement: "SPDXRef-Package-AAA_Tools"
```

So this package has it own repository :  https://example.com/AAA_tools.git.

Now the people that created this project want to exclude some package from ORT scanning, because these packages are just for some tooling.
To define that, they put in place some exclusion in`.ort.yml `:
```
excludes:
  paths:
  - pattern:  "./external/AAA_tools/**"
    reason:   "OTHER"
    comment:  "Files used for configuration of tools."
```

Now what we expect is that, when we run ORT with `--skip-excluded` flag, these excluded packages are not scanned at all i.e. there are not passed to `scanPackages`.
However notice the way the exclusion is done : it is not based on repo url but rather on directory path of the package  (`packageFileName` above).

This PR
1. allows to exclude package dependencies. Currently only exclusion at the project level seemed to be handled.
2. puts in place path-based exclusion for the dependencies. And this is the real difficulty of the thing because a package reference does not contain the original path of the package, therefore we need to do a lookup in the curated packages map.

On the side note, this PR is only the first step.
Afterwards, we need to replace the PathExclude by the FileMatcher. because the glob matching of PathExclude has some issues : when excluding `path`, the pattern **^path$** is used therefore `path/` is not excluded.
Identically, when excluding `path\**` the pattern is **^path/.*$** and thus `path` is not excluded.

**Comment by nnobelis** [comment]
@mnonnenmacher : does my description make sense for you ?

**Comment by mnonnenmacher** [comment]
@nnobelis To make sure I understand the scenario correctly:

- You have a project defined by an spdx.yml in the root directory which has a dependency on "SPDXRef-Package-AAA_Tools".
- "SPDXRef-Package-AAA_Tools" is located in the same source tree in the directory "external/AAA_tools", but it's in a separate repository, so is it included as a Git submodule?
- Does "SPDXRef-Package-AAA_Tools" have an spdx.yml as well?
- You need a way to exclude this package.

I have not worked with projects defined by spdx.yml so far, maybe that's why it's difficult for me to understand the issue. @fviernau Maybe you can help here?

**Comment by nnobelis** [comment]
> @nnobelis To make sure I understand the scenario correctly:
> 
>     * You have a project defined by an spdx.yml in the root directory which has a dependency on "SPDXRef-Package-AAA_Tools".
>     * "SPDXRef-Package-AAA_Tools" is located in the same source tree in the directory "external/AAA_tools", but it's in a separate repository, so is it included as a Git submodule?

Correct. I forgot to give an excerpt of `.gitmodules` :
```
[submodule "external/AAA_tools"]
	path = external/AAA_tools
	url = https://example.com/AAA_tools.git
```

> 
>     * Does "SPDXRef-Package-AAA_Tools" have an spdx.yml as well?

In my case no but I don't think this is relevant to the issue.
> 
>     * You need a way to exclude this package.

Exactly. And in the end, the functionality of this PR is to be able to exclude the dependency  not just by scope or by name, but also by directory path, for convenience.

**Comment by nnobelis** [comment]
Maybe I can split the commit in two :

1. exclude the packages if they are scope excluded in the Scanner if --skip-excluded is specified
2. do the same if they are path excluded

**Comment by nnobelis** [comment]
Work on this branch has been cancelled in favor of project scanner and FossID ignore rules.

**Pull request closed without merging**